code/obj/railing.dm

```diff
@@ -0,0 +1,344 @@
+
+	attack_hand(mob/user)
+		if (railing_is_broken(src))
+			user.show_text("[src] is broken! All you can really do is break it down...", "red")
+		else
+			actions.start(new /datum/action/bar/icon/railing_jump(user, src), user)
+
+	orange
+		color = "#ff7b00"
+
+	red
+		color = "#ff0000"
+
+	green
+		color = "#09ff00"
+
+	yellow
+		color = "#ffe600"
+
+	cyan
+		color = "#00f7ff"
+
+	purple
+		color = "#cc00ff"
+
+	blue
+		color = "#0026ff"
+
+
+/datum/action/bar/icon/railing_jump
+	duration = 1 SECOND
+	interrupt_flags = INTERRUPT_MOVE | INTERRUPT_ACT | INTERRUPT_STUNNED | INTERRUPT_ACTION
+	id = "railing_deconstruct"
+	icon = 'icons/ui/actions.dmi'
+	icon_state = "railing_jump"
+	var/mob/ownerMob
+	var/obj/railing/the_railing
+	var/turf/jump_target //where the mob will move to when they complete the jump!
+	var/is_athletic_jump //if the user has the athletic trait, and therefore does the BEEG HARDCORE PARKOUR YUMP
+	var/no_no_zone //if the user is trying to jump over railing onto somewhere they couldn't otherwise move through...
+
+	New(The_Owner, The_Railing)
+		..()
+		if (The_Owner)
+			owner = The_Owner
+			ownerMob = The_Owner
+			if (ishuman(owner))
+				var/mob/living/carbon/human/H = owner
+				if (H.traitHolder.hasTrait("athletic"))
+					duration = round(duration / 2)
+					is_athletic_jump = 1
+		if (The_Railing)
+			the_railing = The_Railing
+			if (get_dist(ownerMob, the_railing) == 0)
+				jump_target = get_step(the_railing, the_railing.dir)
+			else
+				jump_target = get_turf(the_railing)
+
+	onUpdate()
+		..()
+		// you gotta hold still to jump!
+		if (get_dist(ownerMob, the_railing) > 1)
+			interrupt(INTERRUPT_ALWAYS)
+			ownerMob.show_text("Your jump was interrupted!", "red")
+			return
+
+	onStart()
+		..()
+		if (get_dist(ownerMob, the_railing) > 1 || the_railing == null || ownerMob == null)
+			interrupt(INTERRUPT_ALWAYS)
+			return
+		for(var/mob/O in AIviewers(ownerMob))
+			O.show_text("[ownerMob] begins to pull [himself_or_herself(ownerMob)] over [the_railing].", "red")
+
+	onEnd()
+		..()
+		var/bunp //the name of the thing we have bunp'd into when trying to jump the railing
+		var/list/bunp_whitelist = list(/obj/railing, /obj/decal/stage_edge) // things that we cannot bunp into
+		if (jump_target.density)
+			bunp = jump_target.name
+			no_no_zone = 1
+		else
+			for (var/obj/o in jump_target.contents)
+				for (var/i=1, i<=bunp_whitelist.len+1, i++) //+1 so we can actually go past the whitelist len check
+					if (!(i > bunp_whitelist.len))
+						// if it's an exception to the bunp rule...
+						if (istype (o, bunp_whitelist[i]))
+							break
+
+					// don't proceed just yet if we aren't done going through the whitelist!
+					else if (i <= bunp_whitelist.len)
+						continue
+
+					// otherwise, if it's a dense thing...
+					else if (o.density)
+						bunp = o.name
+						no_no_zone = 1
+						break
```

Comment:
I appreciate that you're trying to make this nice and modular but considering that the whitelist comes with just two items now I'd probably go with just dumb `if(o.density && !istype(o, /obj/railing) && !istype(o, /obj/decal/stage_edge))` instead. But it's not terrible as is so take this only as a suggestion.

Reply:
It was like that originally, but I made it a list so that it's easier to expand the list of things that you can railing-jump onto later.